pkgs/dart_model/lib/src/json_changes.dart

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'json.dart';
+
+/// The changes between two [Json] instances.
+extension type JsonChanges.fromJson(Map<String, Object?> node) {
+  JsonChanges({
+    required List<Update> updates,
+    required List<Removal> removals,
+  }) : this.fromJson({
+          'updates': updates,
+          'removals': removals,
+        });
+
+  /// The nodes that were updated by these changes.
+  List<Update> get updates => (node['updates']! as List).cast();
+
+  /// The nodes that were removed by these changes.
+  List<Removal> get removals => (node['removals']! as List).cast();
+}
+
+/// A path into JSON data.
+extension type Path.fromJson(List<Object?> node) {
+  Path(List<String> path) : this.fromJson(path);
+
+  List<String> get path => (node as List).cast();
+
+  String? get uri => path.isEmpty ? null : path.first;
+
+  /// This path followed by [element].
+  Path followedByOne(String element) =>
+      Path(path.followedBy([element]).toList());
+
+  /// This path except the first entry.
+  Path skipOne() => Path(path.skip(1).toList());
```

Comment:
If Path was an extension type on iterable then at least this could be much faster and avoid the copy. We could also probably avoid a lot of the other list copying by removing the `toList()` calls. Serializing to JSON would still be slow though with tons of duplicate logic in there.

There is in general way to much copying of lists going on in this class imo and a different approach is needed.

Reply:
Paths are gone following the change to maps :)